apps/emqx/src/emqx_session.erl

```diff
@@ -831,115 +458,100 @@ run_terminate_hooks(ClientInfo, takenover, Session) ->
-%% Next Packet Id
-%%--------------------------------------------------------------------
+-spec info(t()) -> emqx_types:infos().
+info(Session) ->
+    maps:from_list(info(?INFO_KEYS, Session)).
 
-obtain_next_pkt_id(Session) ->
-    {Session#session.next_pkt_id, next_pkt_id(Session)}.
+-spec info
+    ([atom()], t()) -> [{atom(), _Value}];
+    (atom(), t()) -> _Value.
+info(Keys, Session) when is_list(Keys) ->
+    [{Key, info(Key, Session)} || Key <- Keys];
+info(impl, Session) ->
+    get_impl_mod(Session);
+info(Key, Session) ->
+    ?IMPL(Session):info(Key, Session).
 
-next_pkt_id(Session = #session{next_pkt_id = ?MAX_PACKET_ID}) ->
-    Session#session{next_pkt_id = 1};
-next_pkt_id(Session = #session{next_pkt_id = Id}) ->
-    Session#session{next_pkt_id = Id + 1}.
+-spec stats(t()) -> emqx_types:stats().
+stats(Session) ->
+    ?IMPL(Session):stats(Session).
 
 %%--------------------------------------------------------------------
-%% Message Latency Stats
+%% Common message events
 %%--------------------------------------------------------------------
-on_delivery_completed(
-    Msg,
-    #session{created_at = CreateAt, clientid = ClientId}
-) ->
+
+on_delivery_completed(Msg, #{clientid := ClientId}, Session) ->
     emqx:run_hook(
         'delivery.completed',
         [
             Msg,
-            #{session_birth_time => CreateAt, clientid => ClientId}
+            #{
+                session_birth_time => ?IMPL(Session):info(created_at, Session),
+                clientid => ClientId
+            }
         ]
     ).
 
-mark_begin_deliver(Msg) ->
-    emqx_message:set_header(deliver_begin_at, erlang:system_time(millisecond), Msg).
+on_replies_delivery_completed(Replies, ClientInfo, Session) ->
+    lists:foreach(
+        fun({_PacketId, Msg}) ->
+            case Msg of
+                #message{qos = ?QOS_0} ->
+                    on_delivery_completed(Msg, ClientInfo, Session);
+                _ ->
+                    ok
+            end
+        end,
+        Replies
+    ).
 
-%%--------------------------------------------------------------------
-%% Helper functions
-%%--------------------------------------------------------------------
+on_dropped_qos2_msg(PacketId, Msg, RC) ->
+    ?SLOG(
+        warning,
+        #{
+            msg => "dropped_qos2_packet",
+            reason => emqx_reason_codes:name(RC),
+            packet_id => PacketId
+        },
+        #{topic => Msg#message.topic}
+    ),
+    ok = emqx_metrics:inc('messages.dropped'),
+    ok = emqx_hooks:run('message.dropped', [Msg, #{node => node()}, emqx_reason_codes:name(RC)]),
+    ok.
 
--compile({inline, [sort_fun/2, batch_n/1, with_ts/1, age/2]}).
+%%--------------------------------------------------------------------
 
-sort_fun({_, A}, {_, B}) ->
-    A#inflight_data.timestamp =< B#inflight_data.timestamp.
+-spec should_discard(message() | emqx_types:deliver()) -> boolean().
+should_discard(MsgDeliver) ->
+    is_banned_msg(MsgDeliver).
```

Comment:
Assuming the logic of `is_banned_msg` is correct, and since this is the only fn used in `should_discard`, it seems like this ought to be:

```suggestion
-spec should_keep(message() | emqx_types:deliver()) -> boolean().
should_keep(MsgDeliver) ->
    is_not_banned_msg(MsgDeliver).
```